src/validation.cpp

```diff
@@ -2502,12 +2515,20 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
             // to abandon a transaction and then have it inadvertently cleared by
             // the notification that the conflicted transaction was evicted.
 
+            // throw all transactions though the signal-interface
+            for (const PerBlockConnectTrace& trace : connectTrace.GetBlocksConnected()) {
+                assert(trace.conflictedTxs);
+                for (const auto& tx : *trace.conflictedTxs) {
+                    GetMainSignals().SyncTransaction(*tx, NULL, CMainSignals::SYNC_TRANSACTION_NOT_IN_BLOCK);
+                }
+            }
+
             // Transactions in the connected block are notified
-            for (const auto& pair : connectTrace.GetBlocksConnected()) {
-                assert(pair.second);
-                const CBlock& block = *(pair.second);
+            for (const PerBlockConnectTrace& trace : connectTrace.GetBlocksConnected()) {
```

Comment:
OK, I'll just squash a few commits before merge, I dont particularly feel strongly about there being so many commits and dont want to have to reorder commits to fix this.